Notifying vSphere Cloud Provider when Node is added or removed.

rjog opened the pull request:
VSphere Cloud Provider (VCP) has requests to support Kubernetes cluster spanning across multiple vSphere datacenters, vCenters. Let's call it as multi-vc support in VCP.

For multi-vc environment VCP needs to undergo large amount of changes.
First change is to get notified when any new kubernetes node VM is added or removed so that VCP can maintain the cache of VM objects correctly.
Based on the various field in the Node object, VCP can determine correct VM in the vSphere inventory.

This change allows VCP to get notified upon node registration / unregistration.
Keeping the change small enough so that it becomes easier for review.
In coming days changes will be posted for review to make use of these notifications in VCP effectively.

This fixes issue: https://github.com/vmware/kubernetes/issues/236

Testing Done:
Verified that with any Kubernetes Node addition / deletion notification functions get called.

rohitjogvmw commented:
release-note-none

dims commented:
/ok-to-test

rohitjogvmw commented:
/test pull-kubernetes-unit
/test pull-kubernetes-verify
/test pull-kubernetes-kubemark-e2e-gce
/test pull-kubernetes-bazel-test
/test pull-kubernetes-e2e-gce-bazel

rohitjogvmw commented:
/retest

Commit:
VSphere Cloud Provider (VCP) has requests to support Kubernetes cluster spanning across multiple vSphere datacenters, vCenters. Let's call

For multi-vc environment VCP needs to undergo large amount of changes.
First change is to get notified when any new kubernetes node VM is added or removed so that VCP can maintain the cache of VM objects corre
Based on the various field in the Node object, VCP can determine correct VM in the vSphere inventory.

This change allows VCP to get notified upon node registration / unregistration.
Keeping the change small enough so that it becomes easier for review.
In coming days changes will be posted for review to make use of these notifications in VCP effectively.

rohitjogvmw commented:
/release-note-none

dims commented:
/area cloudprovider

dims commented:
@rohitjogvmw Can you please run this by the @kubernetes/sig-cluster-lifecycle-feature-requests group? May be the next weekly meeting of sig-cluster-lifecycle as most of the wg-cloudprovider folks are there. Thanks!  

Am asking since you are hinting at larger plans for this interface and it seems to be only useful for one provider at the moment

caesarxuchao commented:
@cheftako could you take a look? Thanks.

caesarxuchao commented:
/unassign @caesarxuchao
/assign @cheftako

jhorwit2 commented:
At first glance this seems like a fairly specific use case. What do you think about using the client builder passed into the `Initialize` call to construct a node informer? Then you could watch for these events without requiring changes to the interface.

andrewsykim commented:
Agreed with @jhorwit2, one reason for out-of-tree providers is so you can add custom logic like this without having to change the core interface. If this can work without adding more methods to cloud interfaces that would be great.

luxas commented:
I don't want to add many more functions to this interface. I want more details and design discussion before we do anything here. To me it seems like this can be implemented in other, better ways.

> First change is to get notified when any new kubernetes node VM is added or removed so that VCP can maintain the cache of VM objects correctly.

Why wouldn't you be able to do this "normally" by periodically LISTing or WATCHing `/api/v1/nodes` from your cloud provider code?

rohitjogvmw commented:
@andrewsykim, @luxas, @jhorwit2
Thanks for your valuable inputs.
We will try to work on the suggested ways.

dims commented:
/unassign @dims

dims commented:
/assign

Let's close this one out please for now.

dims commented:
/close
